pkg/reloader/reloader.go

```diff
@@ -0,0 +1,212 @@
+
+	if err := w.Add(cfgDir); err != nil {
+		return errors.Wrap(err, "add file watch")
+	}
+	if _, err := r.applyConfig(); err != nil {
+		return errors.Wrap(err, "initial apply")
+	}
+	if err := r.triggerReload(ctx); err != nil {
+		level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+	}
+
+	tick := time.NewTicker(3 * time.Minute)
+	defer tick.Stop()
+
+	for {
+		select {
+		case event := <-w.Events:
+			level.Debug(r.logger).Log("msg", "received watch event", "op", event.Op, "name", event.Name)
+
+			if event.Name != r.cfgFilename {
+				continue
+			}
+			changes, err := r.applyConfig()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "apply failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case err := <-w.Errors:
+			level.Error(r.logger).Log("msg", "watch error", "err", err)
+
+		case <-tick.C:
+			changes, err := r.refreshRules()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "refreshing rules failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func hashFile(h hash.Hash, fn string) error {
+	f, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	h.Write([]byte{'\xff'})
+	h.Write([]byte(fn))
+	h.Write([]byte{'\xff'})
+
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reloader) applyConfig() (ok bool, err error) {
+	if r.cfgFilename == "" {
+		return false, nil
+	}
+	h := sha256.New()
+
+	if err := hashFile(h, r.cfgFilename); err != nil {
+		return false, errors.Wrap(err, "hash file")
+	}
+	hb := h.Sum(nil)
+
+	if bytes.Equal(r.lastCfgHash, hb) {
+		return false, nil
+	}
+	r.lastCfgHash = hb
+
+	if r.cfgSubstFilename == "" {
+		return true, nil
+	}
+
+	b, err := envsubst.ReadFileRestricted(r.cfgFilename, true, true)
+	if err != nil {
+		return false, errors.Wrap(err, "read file")
+	}
+	if err := ioutil.WriteFile(r.cfgSubstFilename, b, 0666); err != nil {
+		return false, errors.Wrap(err, "write file")
+	}
+	return true, nil
+}
+
+func (r *Reloader) refreshRules() (bool, error) {
+	if r.ruleDir == "" {
+		return false, nil
+	}
+	h := sha256.New()
+
+	err := filepath.Walk(r.ruleDir, func(path string, _ os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if err := hashFile(h, path); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return false, errors.Wrap(err, "build hash")
+	}
+	hb := h.Sum(nil)
+
+	if bytes.Equal(hb, r.lastRulesHash) {
+		return false, nil
+	}
+	r.lastRulesHash = hb
+
+	return true, nil
+}
+
+func (r *Reloader) triggerReload(ctx context.Context) error {
+	req, err := http.NewRequest("POST", r.promURL.String(), nil)
```

Comment:
what exactly endpoint we want to hit? Shouldn't we hardcode it here? Then variable name: `r.promURL` will make sense.

Reply:
Below in the file is a helper that gives you the default reload endpoint from a Prometheus base URL. We are using that in main. Seems cleaner than to hardcode it deep down here.

Reply:
Cool make sense, but we need here then `reloadPromURL` instead `promURL`

Reply:
still valid